Add tests for referenceTarget interaction with aria-controls and aria-describedby

alice opened the pull request:
#### 7a69959f4c523e16cefed7b38e19a6dccf7bd99f
<pre>
Add tests for referenceTarget interaction with aria-controls and aria-describedby
<a href="https://bugs.webkit.org/show_bug.cgi?id=290601">https://bugs.webkit.org/show_bug.cgi?id=290601</a>

Reviewed by Tyler Wilcock.

* LayoutTests/accessibility/shadow-dom/reference-target/mac/aria-controls-expected.txt: Added.
* LayoutTests/accessibility/shadow-dom/reference-target/mac/aria-controls.html: Added.
* LayoutTests/accessibility/shadow-dom/reference-target/mac/aria-describedby-expected.txt: Added.
* LayoutTests/accessibility/shadow-dom/reference-target/mac/aria-describedby.html: Added.

Canonical link: <a href="https://commits.webkit.org/293557@main">https://commits.webkit.org/293557@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2550d7ef13bb8d3da87767e8083090f43dbe6f3a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98966 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18603 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8847 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104094 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49557 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18898 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27053 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75337 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32464 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101970 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14360 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89379 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55698 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14152 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7354 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48935 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84084 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7429 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106461 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26063 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18998 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84299 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26438 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85576 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83801 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21316 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28462 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6134 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/19787 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26016 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25836 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29156 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27410 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/shadow-dom/reference-target/mac/aria-describedby.html:
```diff
@@ -0,0 +1,34 @@
+
```
Extra newline here I think, and in general there are a few extra newlines before and after a few tags (`html`, `head`, `body`, `</script>`) in this file that we don't typically go for in our tests.

alice replied:
Done.

twilco commented on LayoutTests/accessibility/shadow-dom/reference-target/mac/aria-describedby.html:
```diff
@@ -0,0 +1,34 @@
+let output = expect("input1.customContent", "'description: Description 1'");
+
+debug(output);
+finishJSTest();
```
Ditto [my comment in your other PR](https://github.com/WebKit/WebKit/pull/43301#discussion_r2034148897) about not needing this 🙂

alice replied:
Done, thank you!

twilco commented on LayoutTests/accessibility/shadow-dom/reference-target/mac/aria-controls.html:
```diff
@@ -0,0 +1,54 @@
+<!DOCTYPE HTML>
+<html>
+
```
A few extra newlines in between tags here too. I don't think there's a hard rule here, but in many other accessibility tests we only have empty newlines between "significant" sections of elements in the `<body>` (e.g. I'd say the ones between `<body>` and `<x-panel-2>` are good here).

alice replied:
Did my best to clean up the newlines, PTAL :)

twilco commented on LayoutTests/accessibility/shadow-dom/reference-target/mac/aria-controls.html:
```diff
@@ -0,0 +1,54 @@
+output += expect("tab2.isSelected", "true");
+
+debug(output);
+finishJSTest();
```
Unlike the other cases, this test does actually need to be async. WebKit accessibility has two "modes". The first, and the one that runs in EWS as part of pull requests, is "main-thread mode" — the accessibility objects living on the main-thread get their data directly from the render tree and DOM.

The second is "isolated tree mode", which runs [tests as part of the post-commit process](https://build.webkit.org/#/builders/1231) (this bot is not in a good state at the moment...but usually is). You can run tests in this mode if you are developing on macOS (the only platform it's enabled for at the moment) by passing the `--accessibility-isolated-tree` flag to `run-webkit-tests`.

In this mode, accessibility requests are served from a dedicated accessibility thread, and it takes time for the DOM changes to sync the appropriate accessibility tree updates to this thread (somewhat similar to Chrome's model which I know you are very familiar with).

So long story short, we need to make this test async-friendly in a few ways:

1. Add `window.jsTestIsAsync = true;` after `description`
2. After we focus `panel2Item`, we'll need to use `output += await expectAsync("tab2.isSelected", "true");` instead of `expect`, as it is likely what is written now will query the accessibility tree before the appropriate change is synced.
3. Afterwards, this test should pass with and without the `--accessibility-isolated-tree` flag.

alice replied:
This is really helpful, thank you!

I've put the test body inside an async IIFE, since I don't think `await` will work at the top-level in non-module code. It seems like there a few different patterns used for this; are there any that are generally preferred nowadays?

twilco replied:
The pattern most often used in the accessibility tests is this one with `setTimeout`:

https://github.com/WebKit/WebKit/blob/43dfb26178e86e86570b7a53003e2514ca6fac70/LayoutTests/accessibility/dynamic-speak-as.html#L22

But I think what you have here is fine too.

alice committed:
Add tests for referenceTarget interaction with aria-controls and aria-describedby
https://bugs.webkit.org/show_bug.cgi?id=290601

Reviewed by Tyler Wilcock.

* LayoutTests/accessibility/shadow-dom/reference-target/mac/aria-controls-expected.txt: Added.
* LayoutTests/accessibility/shadow-dom/reference-target/mac/aria-controls.html: Added.
* LayoutTests/accessibility/shadow-dom/reference-target/mac/aria-describedby-expected.txt: Added.
* LayoutTests/accessibility/shadow-dom/reference-target/mac/aria-describedby.html: Added.

Canonical link: https://commits.webkit.org/293557@main
